includes/class-sensei-enrolment-calculation-scheduler.php

```diff
@@ -0,0 +1,137 @@
+<?php
+/**
+ * File containing the class Sensei_Enrolment_Calculation_Scheduler.
+ *
+ * @package sensei
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * The Sensei_Enrolment_Calculation_Scheduler is responsible for running jobs of user enrolment calculations. It is set
+ * up to run once after a Sensei version upgrade or when Sensei_Course_Enrolment_Manager::get_site_salt() is updated.
+ */
+class Sensei_Enrolment_Calculation_Scheduler {
+	const CALCULATION_VERSION_OPTION_NAME = 'sensei-scheduler-calculation-version';
+
+	/**
+	 * Number of seconds between each job run.
+	 *
+	 * @var integer
+	 */
+	private $interval;
+
+	/**
+	 * Number of users for each job run.
+	 *
+	 * @var integer
+	 */
+	private $batch_size;
+
+
+	/**
+	 * Sensei_Enrolment_Calculation_Scheduler constructor.
+	 *
+	 * @param integer $interval The scheduler's interval.
+	 * @param integer $batch_size The scheduler's batch size.
+	 */
+	public function __construct( $interval, $batch_size ) {
+		$this->interval   = $interval;
+		$this->batch_size = $batch_size;
+
+		// phpcs:ignore WordPress.WP.CronInterval.ChangeDetected -- Job size small.
+		add_filter( 'cron_schedules', [ $this, 'add_schedule' ] );
+		add_action( 'sensei_calculate_enrolments', [ $this, 'calculate_enrolments' ], 10, 0 );
+		add_action( 'init', [ $this, 'start' ], 101 );
+	}
+
+	/**
+	 * This method starts the scheduler if the run is not completed for the current Sensei version.
+	 */
+	public function start() {
+		if ( get_option( self::CALCULATION_VERSION_OPTION_NAME ) === Sensei_Course_Enrolment_Manager::get_enrolment_calculation_version() ) {
+			return;
+		}
+
+		if ( ! wp_next_scheduled( 'sensei_calculate_enrolments' ) ) {
+			wp_schedule_event( time(), 'sensei_user_recalculation', 'sensei_calculate_enrolments' );
+		}
+	}
+
+	/**
+	 * Stops the scheduler.
+	 */
+	public function stop() {
+		wp_clear_scheduled_hook( 'sensei_calculate_enrolments' );
+	}
+
+
+	/**
+	 * Marks the run as completed and stops the scheduler.
+	 */
+	public function complete() {
+		update_option(
+			self::CALCULATION_VERSION_OPTION_NAME,
+			Sensei_Course_Enrolment_Manager::get_enrolment_calculation_version()
+		);
+		$this->stop();
+	}
+
+	/**
+	 * Adds a custom WP schedule.
+	 *
+	 * @param array $schedules The existing schedules.
+	 *
+	 * @access private
+	 *
+	 * @return array The schedules
+	 */
+	public function add_schedule( $schedules ) {
+		$schedules['sensei_user_recalculation'] = [
+			'interval' => $this->interval,
+			'display'  => esc_html__( 'Every ' . $this->interval . ' seconds.', 'sensei-lms' ), // phpcs:ignore WordPress.WP.I18n.NonSingularStringLiteralText -- This is unavoidable if we want to use a configurable interval.
```

Comment:
For this rule violation, it probably wants you to use a placeholder.
```suggestion
			// translators: Placeholder %d is interval in seconds.
			'display'  => sprintf( esc_html__( 'Every %d seconds.', 'sensei-lms' ), $this->interval ),
```

Reply:
Fixed 313886990129312d4e86caa929cc4c87a3c6f6f4